### PR TITLE
Fix forward references in ExpressionBuilder

### LINKED ISSUE #397: Solve to fix ExpressionBuilder when definitions are not in-order
Predecessor issue #422 

* `ExpressionBuilderContext.InvokeDefinedFunctionThroughRuntimeContext` when definitions are out of order
* Fix unit tests - Find the `[Ignore("Will fix in https://github.com/FirelyTeam/firely-cql-sdk/issues/397")]` attributes from 4 unit tests and fix them

Check additional todo's  in Slack discussion : https://firelyteam.slack.com/archives/C05GW1J1HFV/p1721036253203929

### LINKED ISSUE #410: Start using FunctionRef.signature to resolve overloads
The Java stack (and us) now produce signatures, so we can simplify the overload resolution logic in the ExpressionBuilder.

### LINKED ISSUE #442: Tuple types with different element ordres being treated as distinct types
```
library ListTest version '1.0.0'

define private L: { Tuple { b: 2, a: 1 }, Tuple { a: 1, b: 2 } }
```

This type of `L` is `List<System.Any>` because we are treating these tuples as different types.

### LINKED ISSUE #443: Retrieves in context unfiltered are being generated with empty, non-null ExpressionRefs
```
library Bug version '1.0.0'

using FHIR version '4.0.1'

define patients: [Patient]
```
The resulting `Retrieve` should have a `null` `context` property.

### PR DESCRIPTION
**Pull request by ewoutkramer**
Fixes #397.  Fixes #410.

This PR enables the translation of FunctionRef/ExpressionRef to rely on the ref's `resultTypeSpecifier` instead of having to look the definition up in the DefinitionDictionary. Not using the DefinitionDictionary avoids running into the situation where the ref is a forwards reference whereby the DefinitionDictionary does not yet contain the translated function we want to call.

This PR uses the recent functionality added to LibrarySet to lookup a symbol, and expands this to be able to lookup functions by signature. Given the function name in the FUnctionRef/ExpressionRef, and, in the case of an overload, the signature, we can now lookup any function defined in any library in the LibrarySet, so we can copy over the resultTypeSpecifier.

Filling out this resultTypeSpecifier is now done in a pre-processing step, happening when we are constructing the LibrarySet. The pre-processing step will find all Refs, and then do this:

* Try to find the function by name. If it is not an overload, use the resultTypeSpecifier
* If there is an overload, use the Ref's signature, if available.
* If there is no signature, use the types from the Ref's parameters and match on those.

Otherwise, we cannot find the overload, and we will raise an error. This ensures that after the pre-processor has run, *all* Refs have a resultTypeSpecifier, and we can depend on that fact.

This PR re-uses the OverloadedFunctionDefinition, and there are some updates to the `CanCombine` function to make sure the pre-processor fix and the CanCombine use the same criteria.

I also added a bunch of useful extension methods for working with the Elm function elements and signatures.

### CODE REVIEW

File: Cql/Elm/ElmTreeWalker.cs [hunk 1]
```diff
@@ -14,97 +14,100 @@
...
+        protected override bool IsRelevant(Type nodeType) => filter?.Invoke(nodeType) ?? base.IsRelevant(nodeType);
+
+        protected override bool Process(object node) => visitor(node);
+    }
+
     /// <summary>
     /// A walker that does a depth-first traversal of a tree and calls a visitor function for each node.
     /// </summary>
-    internal class ElmTreeWalker
+    internal abstract class BaseElmTreeWalker
```
Comment (baseTwo) [inline Cql/Elm/ElmTreeWalker.cs]: This class is significant enough to deserve its own file

File: Cql/Elm/ElmTreeWalker.cs [hunk 2]
```diff
@@ -14,97 +14,100 @@
...
+            if(!IsRelevant(node.GetType()))
+                return;
+
+            if (_visited.Any(i => ReferenceEquals(i, node)))
+                throw new InvalidOperationException("Cycle in ELM graph detected.");
 
-            var properties = root.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance)
-                .Where(p => p.CanRead && isRelevant(p) && p.GetValue(root) is { } propValue)
-                .ToArray();
+            if (Process(node)) return;
```
Comment (baseTwo) [inline Cql/Elm/ElmTreeWalker.cs]: Return on its own line, followed by blank line

File: Cql/Elm/Errors.cs [hunk 3]
```diff
@@ -32,4 +32,14 @@ internal interface ILibraryError : ICqlError
 internal readonly record struct MissingAliasError(IGetLibraryName Source) : ICqlError
 {
     public string GetMessage() => $"{Source.GetType().Name} did not have an alias. Source: {Source}";
+}
+
+internal readonly record struct UntypedOperandInFunctionError(IFunctionElement Func, OperandDef Element) : ICqlError
+{
+    public string GetMessage() => $"Function {Func.Name}' has a parameter '{Element.name}' without a type specifier.";
```
Comment (baseTwo) [inline Cql/Elm/Errors.cs]: Missing ' before {Func.Name}

File: Cql/Elm/Errors.cs [hunk 4]
```diff
@@ -32,4 +32,14 @@ internal interface ILibraryError : ICqlError
...
+}
+
+internal readonly record struct UntypedOperandInFunctionRefError(FunctionRef Func, int Index) : ICqlError
+{
+    public string GetMessage() => $"Function reference to {Func.name}', and parameter {Index} has no type specifier.";
```
Comment (baseTwo) [inline Cql/Elm/Errors.cs]: Missing ' before {Func.name}

File: Cql/Elm/ElmTreeWalker.cs [hunk 5]
```diff
@@ -14,97 +14,100 @@
 
 namespace Hl7.Cql.Elm
 {
+    /// <summary>
+    /// Construct a walker.
```
Comment (baseTwo) [inline Cql/Elm/ElmTreeWalker.cs]: It isn't constructing the walker , it _is_ the walker. The description needs to be a more detailed

File: Cql/Elm/ElmTreeWalker.cs [hunk 6]
```diff
@@ -14,97 +14,100 @@
...
+    /// </summary>
+    /// <param name="visitor">A function that is passed the nodes in the tree. It should return <c>false</c> if this function
+    /// handles the children of the nodes by itself, or <c>true</c> if the walker should visit the children automatically.</param>
+    /// <param name="filter">A function that controls whether the walker visits a child property or not.</param>
```
Comment (baseTwo) [inline Cql/Elm/ElmTreeWalker.cs]: From the description it mentions a predicate against a property, however, the predicate takes a Type. One would have to assume its the type of the value of the property, but this needs to be more explicitly explained

File: Cql/Cql.Compiler/LibrarySetExpressionBuilderContext.cs [hunk 7]
```diff
@@ -31,8 +33,15 @@ internal partial class LibrarySetExpressionBuilderContext
     public DefinitionDictionary<LambdaExpression> ProcessLibrarySet() =>
         this.CatchRethrowExpressionBuildingException(_ =>
         {
+            var ambiguousOverloadCorrector = new AmbiguousOverloadCorrector();
+
             foreach (var library in LibrarySet)
             {
+                // Make sure all overloads in the library are unique.
+                // This is a fix for QICore-based CQL, where the functions only differ by profiles on the same resource.
+                // We should remove this when the compiler is fixed.
```
Comment (baseTwo) [inline Cql/Cql.Compiler/LibrarySetExpressionBuilderContext.cs:42]: What is the issue number for this todo?

File: Cql/Cql.Compiler/ExpressionBuilderContext.StaticHelpers.cs [hunk 8]
```diff
@@ -51,36 +51,31 @@ private static bool TryCorrectQiCoreBindingError(Type source, Type to, out Type?
...
-            10 => typeof(Func<,,,,,,,,,>).MakeGenericType(funcTypeParameters),
-            11 => typeof(Func<,,,,,,,,,,>).MakeGenericType(funcTypeParameters),
-            12 => typeof(Func<,,,,,,,,,,,>).MakeGenericType(funcTypeParameters),
-            13 => typeof(Func<,,,,,,,,,,,,>).MakeGenericType(funcTypeParameters),
-            14 => typeof(Func<,,,,,,,,,,,,,>).MakeGenericType(funcTypeParameters),
-            15 => typeof(Func<,,,,,,,,,,,,,,>).MakeGenericType(funcTypeParameters),
-            _ => throw new NotSupportedException("Functions with more than 15 parameters are not supported.")
-        };
+        Expression.GetFuncType(funcTypeParameters);
+        // funcTypeParameters.Length switch
```
Comment (baseTwo) [inline Cql/Cql.Compiler/ExpressionBuilderContext.StaticHelpers.cs]: Commented out code can be removed

File: Cql/Cql.Compiler/AmbiguousOverloadCorrector.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+using Hl7.Cql.Elm;
+
+namespace Hl7.Cql.Compiler;
+
+/// <summary>
+/// Ensures that there are no overloads with exactly the same signature.
+/// </summary>
+/// <remarks>This happens currently in QICore-based CQL, where the functions only differ
+/// by profiles on the same resource, which results in the same signature after the QICore
+/// profile names are erased by the compiler. This is a temporary fix that should be
```
Comment (baseTwo) [inline Cql/Cql.Compiler/AmbiguousOverloadCorrector.cs:21]: Please include the issue number for this todo

File: Cql/Elm/IDefinitionElement.cs [hunk 10]
```diff
@@ -171,23 +173,18 @@ public partial class ExpressionDef : IDefinitionElement
 
         IDefinitionElement IDefinitionElement.AddError(CqlToElmError error) => this.AddError(error);
 
+        public override string ToString() => $" {Name} : {resultTypeSpecifier}";
```
Comment (baseTwo) [inline Cql/Elm/IDefinitionElement.cs]: The space before {Name} can be removed. If there was a reason for that space here, its a code smell.

File: Cql/Elm/IDefinitionElement.cs [hunk 11]
```diff
@@ -412,6 +411,7 @@ public static OverloadedFunctionDef Create(params IFunctionElement[] elements)
 
         public OverloadedFunctionDef Add(FunctionDef function)
         {
+            // TODO: This does not guard against duplicate signatures, like the Create method does.
```
Comment (baseTwo) [inline Cql/Elm/IDefinitionElement.cs:414]: TODOs need an issue number if it is going to be fixed later

**Review by baseTwo — CHANGES_REQUESTED** [review]

**Review by baseTwo — DISMISSED** [review]

**Comment by EvanMachusak** [comment]
I've identified a small issue where we are creating empty `ExpressionRef` nodes for `Retireve.context` for statements that are in `context Unfiltered` and this is causing exceptions during the preprocessing phase.  I will fix it.

**Review by EvanMachusak — DISMISSED** [review]
All tests pass.  All HEDIS passes.

**Review by EvanMachusak — DISMISSED** [review]

**Review by EvanMachusak — APPROVED** [review]

**Comment by baseTwo** [comment]
@EvanMachusak, none of the demo projects are able to rebuild as a result of the additional commits. Could you please have a look? Issue #450

**Comment by baseTwo** [comment]
I'm going to revert the entire previous PR #437. This means the related issues (#397, #410, #442 and #443 ) will be undone and moved back to TODO.

**Comment by baseTwo** [comment]
@ewoutkramer I tried to duplicate the work into a new PR, but still getting stuck
https://github.com/FirelyTeam/firely-cql-sdk/pull/468

**Pull request merged**